### PR TITLE
Events: added support for np.nan

### PR DESCRIPTION
**Pull request by prassepaul**
## Implemented changes

- [ ] added np.nan support for the event detection algos (added np.nanmean, np.nanmax)
- [ ] created test-cases for np.nan in position and velocities

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change is or requires a documentation update

## How Has This Been Tested?

- [ ] added test-cases

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Review by dkrako — CHANGES_REQUESTED** [review]
I think that it's unexpected that a consecutive sequence of `nan` is a positive for a detected event.

After all, maybe it makes more sense to treat `nan` as not belonging to the event?
I would prefer this behavior at least.

**Comment by prassepaul** [comment]
I think I will implement the behavior as discussed in the meeting:
- do not consider nans during the calculation and delete nans in the beginning and at the end
- add a flag to filter events with nans inside (-> create individual events)

**Review by dkrako — CHANGES_REQUESTED** [review]
Thanks for all this work and also writing all these tests!

Regarding the functionality:
I think it would be best to rename `flag_split_at_nan` into `include_nan`.
See my comments for `engbert.py` in L43.

File: src/pymovements/events/engbert.py [hunk 1]
```diff
@@ -96,18 +100,36 @@ def microsaccades(
...
     # If value is greater than 1, point lies outside the ellipse.
     outside_ellipse = np.greater(np.sum(np.power(velocities / radius, 2), axis=1), 1)
 
-    # Get all indices with velocities outside of ellipse.
-    outside_ellipse_indices = np.where(outside_ellipse)[0]
+    # Get all indices with velocities outside of ellipse or with np.nan.
+    outside_ellipse_indices = np.where(
+        np.logical_or(
+            outside_ellipse,
+            np.isnan(velocities[:, 0]),
```
[the remark this answers is not in the document]
Comment (dkrako) [reply src/pymovements/events/engbert.py:118]: or actually write it like this:
```
np.isnan(velocities).any(axis=1)
```
This doesn't enforce that nans have to be equal on both channels.

File: src/pymovements/events/engbert.py [hunk 2]
```diff
@@ -39,6 +40,7 @@ def microsaccades(
     threshold_factor: float = 6,
     minimum_duration: int = 6,
     minimum_threshold: float = 1e-10,
+    flag_split_at_nan: bool = True,
```
[the remark this answers is not in the document]
Comment (prassepaul) [reply src/pymovements/events/engbert.py:43]: ok. I think you are right. It was from a implementing persepective. I will rename the parameter.

File: src/pymovements/events/engbert.py [hunk 3]
```diff
@@ -96,18 +100,36 @@ def microsaccades(
...
     # Radius of elliptic threshold.
     radius = threshold * threshold_factor
 
     # If value is greater than 1, point lies outside the ellipse.
     outside_ellipse = np.greater(np.sum(np.power(velocities / radius, 2), axis=1), 1)
 
-    # Get all indices with velocities outside of ellipse.
-    outside_ellipse_indices = np.where(outside_ellipse)[0]
+    # Get all indices with velocities outside of ellipse or with np.nan.
+    outside_ellipse_indices = np.where(
```
[the remark this answers is not in the document]
Comment (prassepaul) [reply src/pymovements/events/engbert.py:115]: you are right. I changed it in the code

File: src/pymovements/events/engbert.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (prassepaul) [reply src/pymovements/events/engbert.py:118]: added check for both channels

File: src/pymovements/events/engbert.py [hunk 4]
```diff
@@ -96,18 +100,36 @@ def microsaccades(
             f' ({threshold} < {minimum_threshold})',
         )
 
+    # check if np.nan is in data
```
[the remark this answers is not in the document]
Comment (prassepaul) [reply src/pymovements/events/engbert.py:103]: I think this is no problem any more.

File: src/pymovements/events/engbert.py [hunk 5]
```diff
@@ -96,18 +100,36 @@ def microsaccades(
...
+            outside_ellipse,
+            np.isnan(velocities[:, 0]),
+        ),
+    )[0]
 
     # Get all saccade candidates by grouping all consecutive indices.
     candidates = consecutive(arr=outside_ellipse_indices)
 
+    # Filter np.nan in candidates (delete starting/ending np.nans)
+    if flag_contains_nans:
```
[the remark this answers is not in the document]
Comment (prassepaul) [reply src/pymovements/events/engbert.py:126]: done.

File: src/pymovements/events/ivt.py [hunk 6]
```diff
@@ -29,13 +29,15 @@
...
 from pymovements.utils.checks import check_shapes_positions_velocities
+from pymovements.utils.filters import filter_and_split
 
 
 def ivt(
         positions: list[list[float]] | list[tuple[float, float]] | np.ndarray,
         velocities: list[list[float]] | list[tuple[float, float]] | np.ndarray,
         velocity_threshold: float,
         minimum_duration: int,
+        flag_split_at_nan: bool = True,
```
[the remark this answers is not in the document]
Comment (prassepaul) [reply src/pymovements/events/ivt.py:40]: done

File: src/pymovements/events/idt.py [hunk 7]
```diff
@@ -46,14 +48,15 @@ def dispersion(positions: list[list[float]] | np.ndarray) -> float:
...
-    return sum(np.max(positions, axis=0) - np.min(positions, axis=0))
+    return sum(np.nanmax(positions, axis=0) - np.nanmin(positions, axis=0))
 
 
 def idt(
         positions: list[list[float]] | list[tuple[float, float]] | np.ndarray,
         velocities: list[list[float]] | list[tuple[float, float]] | np.ndarray,
         dispersion_threshold: float,
         minimum_duration: int,
+        flag_split_at_nan: bool = True,
```
[the remark this answers is not in the document]
Comment (prassepaul) [reply src/pymovements/events/idt.py:59]: done

**Review by dkrako — CHANGES_REQUESTED** [review]
Final minor changes requested, won't take much.
Thanks a lot for the work!

**Review by dkrako — APPROVED** [review]
LGTM!

**Pull request merged**